packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -664,6 +658,140 @@ export class RunningSummarizer implements IDisposable {
 		});
 	}
 
+	/**
+	 * Tries to summarize with dynamic retries, i.e., retry is based on the failure params and error received.
+	 * For instance, if summarize failed due to a server error that has `retryAfterSeconds` set, wait for that duration
+	 * and retry summarize. Similarly, if summary generation failed but the error is retriable, retry before bailing out.
+	 */
+	private async trySummarizeWithDynamicRetries(
+		reason: SummarizeReason,
+		cancellationToken = this.cancellationToken,
+	) {
+		let previousStage: SummaryStage | undefined;
+
+		// The maximum number of attempts across failure stages. If summarization does not succeed by these number
+		// of attempts, quit and close the summarizer.
+		const maxAttempts = defaultMaxAttempts;
+		let currentAttempts = 0;
+
+		// The maximum number of attempts for a given stage. The default value can be overridden via failure
+		// params. Each failure is different and some can be succeeded by attempting more often than others.
+		// If summarization fails at different stages, each stage gets some attempts. However, the total number of
+		// attempts across stages cannot exceed `maxAttempts`.
+		let maxAttemptsForStage = defaultMaxAttemptsForStage;
+		let currentAttemptsForStage = 0;
+
+		type ShouldRetryResult =
+			| { retry: false; stage: SummaryStage | undefined }
+			| {
+					retry: true;
+					stage: SummaryStage;
+					delaySeconds: number;
+					stageChanged: boolean;
+					retryCount: number | undefined;
+			  };
+		// Returns whether summarization should be retried based on the failure data.
+		const shouldRetry = (
+			failureData: ISubmitSummaryFailureResult | undefined,
+		): ShouldRetryResult => {
+			// For lastSummary, do not try another attempt.
+			// If the failure data does not specify `retryAfterSeconds` or the stage at which summarization failed,
+			// do not try another attempt.
+			// If maxAttempts have been exceeded, do not try another attempt.
+			if (
+				reason === "lastSummary" ||
+				failureData === undefined ||
+				failureData.retryAfterSeconds === undefined ||
+				failureData.stage === undefined ||
```

Comment:
Probably can remove

Reply:
I reverted most of the changes here. Please take a look.